config-parser: unit tests

logiclrd opened the pull request:
Add unit tests based on the commented-out test code at the end of config-parser.c.
Add a facility for temporary files associated with tests, with autogenerated names and automatic clean-up at the end of the test run. Files are added to a log in a separate file, and that log is later read in to effect clean-up. In this way, a test that creates a temporary file and then crashes does not "forget" about the file it created.

logiclrd committed:
Added test facility for temporary files, automatically cleaned up at the end of the test run:
- Added test/tempfile.c and include/test-tempfile.h. Updated Makefile.am for new files.
- Updated test/harness.c to call test_temp_files_cleanup before exit.
Added test cases based on the commented-out test code at the end of config-parser.c, but updated for the new len semantic in cfg_get_string. Updated Makefile.am for new file. Added new test cases to test-funcs.h.
Updated the comment on the declaration of cfg_get_string for accuracy.

logiclrd committed:
Updated posix_exec to treat CLD_KILLED and CLD_DUMPED as successes, since what's being measured as successful or not is whether the child process ran in the first place, not how it exited.

logiclrd committed:
Added parameter int *abnormal_exit to os_exec, win32_exec and posix_exec. Updated the implementations to assign it a boolean value based on whether the process appears to have exited abnormally.
Updated run_test_child in test/harness.c to check for abnormal exit and translate it to SCHISM_TESTRESULT_CRASH.

logiclrd committed:
Updated schism_test_main in test/harness.c to flush stdout after adding "CRASH" to the end of the status line for a crashed test, otherwise it will likely remain buffered while future tests run and complete.

logiclrd committed:
Oops, I had only updated the win32_exec_[AW] variants. Updated win32_exec to match, taking the abnormal_exit parameter and passing it along to the corresponding variant.
Updated win32_run_hook to include a value (NULL) for the abnormal_exit parameter to win32_exec.
Fixed up test/tempfile.c on Win32:
- Switched test_temp_file from mkstemp to mkfstemp, provided by util.c. Added an include directive for util.h.
- Added a line to redirect unlink to _unlink on Win32 builds.
- Implemented strdup by hand using malloc and strcpy, as strdup per se does not seem to exist on Windows.

mrpapersonic commented on test/tempfile.c:
```diff
@@ -0,0 +1,104 @@
+		remaining = template_length;
+
+		while (1) {
+			int num_written = fwrite(buf, 1, remaining, f);
```
i think `fwrite` can only return less than `remaining` if an error occurs (in which case any further writes will be blocked)

logiclrd replied:
According the documentation for `fwrite`, there are plenty of transitive reasons why a write might be partial:

* Insufficient disk space (could change at any time)
* Interrupted syscalls
* Device-specific limitations ("Some devices or file systems might impose limitations that lead to partial writes in specific scenarios")

When a partial write happens, `fwrite` doesn't return an error. I'm pretty sure it is canonically correct to treat a partial write as a success and retry with the remaining data. If the error was not transient, then the next call will fail to write any items. But, the way the interface is defined, the next call _could_ succeed and write more items.

mrpapersonic commented on test/harness.c:
```diff
@@ -63,10 +64,15 @@ static testresult_t run_test_child(const char *self, test_index_entry *entry)
+
+	if (os_exec(&status, &abnormal_exit, NULL, self, entry->name, (char *)NULL))
+	{
+		if (abnormal_exit)
```
IMO this should be use the ternary operator

logiclrd replied:
Can do.

mrpapersonic commented on test/cases/config-parser.c:
```diff
@@ -0,0 +1,809 @@
+
+#include "config-parser.h"
+
+#define CONFIG_FILE \
```
can this just be a static const global?
```
static const char *sample_config =
    "[ducks]\n"
    <...>
```

logiclrd replied:
I don't see why not. :-)

mrpapersonic commented on test/tempfile.c:
```diff
@@ -0,0 +1,104 @@
+#define unlink _unlink
+#endif
+
+int test_temp_file(char **temp_file, const char *template, int template_length)
```
`temp_file` has a maximum character count anyway (16), so we can just `#define` that value as a constant in `test-tempfile.h` and not have to worry about freeing it in the tests

logiclrd replied:
Fair enough :-)

mrpapersonic commented on test/cases/config-parser.c:
Like I did with baa66af1eed6af175177a49c2f3b3287f8c0cdad, the "common sequence" can be moved into its own function, maybe `static int test_config_file_init(cfg_file_t *cfg)`, and the tests can just call into it.

logiclrd replied:
Done. I used a macro so that it could return directly from its point of invocation.

logiclrd committed:
PR feedback:
- Moved temp file name template into test-tempfile.h. Made a macro for the necessary buffer size. Reworked test_temp_file to take a static buffer. Updated all call sites correspondingly.
- Changed CONFIG_FILE macro in test/cases/config-parser.c into a static buffer test_config_file_content.
- Factored common test arrangement in test/cases/config-parser.c to a macro.
- Updated run_test_child in test/harness.c to use a ternary expression to save a few lines.

mrpapersonic commented on test/cases/config-parser.c:
```diff
@@ -0,0 +1,579 @@
+	// Assert
+	ASSERT(strcmp(colour, "brown") == 0);
+	ASSERT(count == 7);
+	ASSERT(weight == 64);
```
`cfg_free` ? :)

logiclrd replied:
Oh, I forgot that was a thing. :grin:

logiclrd replied:
Hmm, in the absence of a scope-exit structure like `try/finally` or C++ scoped objects, I don't think there's a way to guarantee that it always gets freed properly. But, we can at least have it free it on the happy path. A passing test run thus includes known-working `cfg_free`. Since tests are isolated in their own processes, failed tests are going to immediately have their entire memory space torn down anyway. Acceptable?

logiclrd committed:
Added a cfg_free call to the happy path of every test in test/cases/config-parser.c.

mrpapersonic commented on test/cases/config-parser.c:
```diff
@@ -0,0 +1,617 @@
+	int num_null_section = cfg_get_number(&cfg, NULL, "shouldn't crash", 1);
+	int num_null_key = cfg_get_number(&cfg, "shouldn't crash", NULL, 1);
+	const char *str_null_default_value = cfg_get_string(&cfg, "doesn't", "exist", NULL, 0, NULL);
+
```
as per b11e3dd9ab191f20830062be74f6035af4db78ba we should have a test that supplies a valid buffer but puts in an (obviously bad) 0 for the length
